### DISCUSSION TITLE
Working with GPUs and Cuda

### DISCUSSION THREAD

**Discussion by sdspieg** [explosion/spaCy#12353] [Help: Best practices]
Could somebody please just post instructions for how to get spaCy to work with CUDA on Ubuntu or Windows? CUDA is now at version 12.1. Is it really true that we have to go back to 11.6 to get this to work with pytorch? And should we then do this in a separate virtual environment where we then set the CUDA_HOME environment variable to the path of the older CUDA version? WIll that interfere with the (more up-to-date) host version of CUDA? Are there any other tips or tricks to take maximal advantage of these new GPU beasts? It would be really great for non-experts like myself to have somebody who knows more about this to walk us through this. Thanks!

**Comment by adrianeboyd** [comment]
We realize that installing CUDA and PyTorch can be frustrating, but spacy doesn't have any unusual project-specific requirements. The main things to get installed and working are CUDA, CuPy, and optionally PyTorch for transformer models.

Each project provides better up-to-date install guides than we would be able to maintain:

* CUDA: https://docs.nvidia.com/cuda/cuda-quick-start-guide/index.html
* CuPy: https://docs.cupy.dev/en/stable/install.html
* PyTorch: https://pytorch.org/get-started/locally/

Right now I do think it is easier to stick with CUDA 11.x. CUDA 11.8 should be fine with the current PyTorch/CuPy releases. If you're looking for the CUDA 11.8 links on the nvidia website:
* https://docs.nvidia.com/cuda/archive/11.8.0/
* https://developer.nvidia.com/cuda-11-8-0-download-archive

Our general install advice related to installing transformers + PyTorch: https://spacy.io/usage/embeddings-transformers/#transformers-installation. You typically only need to set `CUDA_PATH` if you have more than one version of CUDA installed in parallel or CUDA is not in a standard location in your path.

**Comment by clemwangds** [comment]
(time passes...)

I'm on Ubuntu 24, Python 3.12.3, Spacy 3.8.0, CUDA 12.9 running on an old intel laptop

The error du jour is now which I don't know how to fix (or even what this means exactly) and Google AI isn't helping:

```
import spacy
spacy.prefer_gpu()
False
nlp = spacy.load("en_core_web_lg")

2025-07-21 11:25:59.403417: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
E0000 00:00:1753133159.419707 1914091 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1753133159.425063 1914091 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1753133159.438745 1914091 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1753133159.438768 1914091 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1753133159.438770 1914091 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1753133159.438773 1914091 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
```

(added)
This isn't a spacy problem.  I get the same problem with tf (version 2.19.0)

```
import tensorflow as tf
2025-07-21 12:57:29.412591: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
E0000 00:00:1753138649.577967    6829 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1753138649.626658    6829 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1753138650.006276    6829 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1753138650.006304    6829 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1753138650.006307    6829 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1753138650.006309    6829 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
```
